File: app/src/main/java/com/example/escaneodematerialeskof/ui/gestion/GestionReinicioActivity.java
package com.example.escaneodematerialeskof.ui.gestion;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.MenuItem;
import android.widget.TextView;
import android.widget.Toast;
import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.lifecycle.ViewModelProvider;
import com.example.escaneodematerialeskof.R;
import com.example.escaneodematerialeskof.viewmodel.InventoryComparisonViewModel;
import com.google.android.material.navigation.NavigationView;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Actividad para gestionar los diferentes tipos de reinicio del sistema
 */
public class GestionReinicioActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {

    private DrawerLayout drawerLayout;
    private NavigationView navigationView;
    private Toolbar toolbar;

    // Vistas del layout
    private TextView tvEstadoSistema;
    private TextView tvEstadoEscaneado;
    private TextView tvEstadoComparacion;
    private TextView tvHistorialReinicios;
    private androidx.cardview.widget.CardView cardReiniciarComparacion;
    private androidx.cardview.widget.CardView cardReiniciarEscaneado;
    private androidx.cardview.widget.CardView cardReiniciarCompleto;

    // ViewModel y SharedPreferences
    private InventoryComparisonViewModel viewModel;
    private SharedPreferences sharedPreferences;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_gestion_reinicio);

        initializeViews();
        setupToolbar();
        setupNavigationDrawer();
        setupViewModel();
        setupClickListeners();
        actualizarEstadoSistema();
        cargarHistorialReinicios();

        // Manejo moderno del botón Atrás
        getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                if (drawerLayout != null && drawerLayout.isDrawerOpen(GravityCompat.START)) {
                    drawerLayout.closeDrawer(GravityCompat.START);
                } else {
                    setEnabled(false);
                    getOnBackPressedDispatcher().onBackPressed();
                }
            }
        });
    }

    private void initializeViews() {
        drawerLayout = findViewById(R.id.drawer_layout);
        navigationView = findViewById(R.id.nav_view);
        toolbar = findViewById(R.id.toolbar);

        tvEstadoSistema = findViewById(R.id.tv_estado_sistema);
        tvEstadoEscaneado = findViewById(R.id.tv_estado_escaneado);
        tvEstadoComparacion = findViewById(R.id.tv_estado_comparacion);
        tvHistorialReinicios = findViewById(R.id.tv_historial_reinicios);

        cardReiniciarComparacion = findViewById(R.id.card_reiniciar_comparacion);
        cardReiniciarEscaneado = findViewById(R.id.card_reiniciar_escaneado);
        cardReiniciarCompleto = findViewById(R.id.card_reiniciar_completo);
    }

    private void setupToolbar() {
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setTitle("Gestión de Reinicio");
        }
    }

    private void setupNavigationDrawer() {
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawerLayout, toolbar,
                android.R.string.ok,
                android.R.string.cancel
        );
        drawerLayout.addDrawerListener(toggle);
        toggle.syncState();
        navigationView.setNavigationItemSelectedListener(this);
    }

    private void setupViewModel() {
        viewModel = new ViewModelProvider(this).get(InventoryComparisonViewModel.class);
        sharedPreferences = getSharedPreferences("reinicio_historial", MODE_PRIVATE);

        // Observar cambios en los datos
        viewModel.getInventarioSistema().observe(this, inventario -> actualizarEstadoSistema());
        viewModel.getInventarioEscaneado().observe(this, inventario -> actualizarEstadoSistema());
        viewModel.getComparacion().observe(this, comparacion -> actualizarEstadoSistema());
    }

    private void setupClickListeners() {
        cardReiniciarComparacion.setOnClickListener(v ->
                mostrarDialogoConfirmacion(
                        "Reiniciar Comparación",
                        "¿Estás seguro de que deseas reiniciar solo la comparación?\n\n" +
                                "Se mantendrán:\n• Inventario del sistema\n• Inventario escaneado\n\n" +
                                "Se eliminará:\n• Resultados de comparación",
                        this::ejecutarReinicioComparacion
                )
        );

        cardReiniciarEscaneado.setOnClickListener(v ->
                mostrarDialogoConfirmacion(
                        "Reiniciar Inventario Escaneado",
                        "¿Estás seguro de que deseas reiniciar el inventario escaneado?\n\n" +
                                "Se mantendrá:\n• Inventario del sistema\n\n" +
                                "Se eliminará:\n• Inventario escaneado\n• Resultados de comparación",
                        this::ejecutarReinicioEscaneado
                )
        );

        cardReiniciarCompleto.setOnClickListener(v ->
                mostrarDialogoConfirmacionCompleto()
        );
    }

    private void actualizarEstadoSistema() {
        // Actualizar estado del inventario del sistema
        if (viewModel.getInventarioSistema().getValue() != null &&
                !viewModel.getInventarioSistema().getValue().isEmpty()) {
            int items = viewModel.getInventarioSistema().getValue().size();
            tvEstadoSistema.setText("🟢 Cargado (" + String.format("%,d", items) + " items)");
            tvEstadoSistema.setTextColor(getResources().getColor(R.color.success_dark, null));
        } else {
            tvEstadoSistema.setText("🔴 Sin cargar");
            tvEstadoSistema.setTextColor(getResources().getColor(R.color.error_dark, null));
        }

        // Actualizar estado del inventario escaneado
        if (viewModel.getInventarioEscaneado().getValue() != null &&
                !viewModel.getInventarioEscaneado().getValue().isEmpty()) {
            int items = viewModel.getInventarioEscaneado().getValue().size();
            tvEstadoEscaneado.setText("🟡 En progreso (" + String.format("%,d", items) + " items)");
            tvEstadoEscaneado.setTextColor(getResources().getColor(R.color.warning_dark, null));
        } else {
            tvEstadoEscaneado.setText("🔴 Sin datos");
            tvEstadoEscaneado.setTextColor(getResources().getColor(R.color.error_dark, null));
        }

        // Actualizar estado de la comparación
        if (viewModel.getComparacion().getValue() != null &&
                !viewModel.getComparacion().getValue().isEmpty()) {
            // Calcular diferencias usando Triple<String, InventarioItem, InventarioItem>
            int diferencias = 0;
            for (var item : viewModel.getComparacion().getValue()) {
                // item es Triple<String, InventarioItem, InventarioItem>
                // item.getFirst() = SKU, item.getSecond() = sistema, item.getThird() = escaneado
                int diferencia = viewModel.calcularDiferencia(item.getSecond(), item.getThird());
                if (diferencia != 0) {
                    diferencias++;
                }
            }
            tvEstadoComparacion.setText("🟢 Completada (" + diferencias + " diferencias)");
            tvEstadoComparacion.setTextColor(getResources().getColor(R.color.success_dark, null));
        } else {
            tvEstadoComparacion.setText("🔴 Sin realizar");
            tvEstadoComparacion.setTextColor(getResources().getColor(R.color.error_dark, null));
        }
    }

    private void mostrarDialogoConfirmacion(String titulo, String mensaje, Runnable accion) {
        new AlertDialog.Builder(this)
                .setTitle(titulo)
                .setMessage(mensaje)
                .setPositiveButton("Confirmar", (dialog, which) -> {
                    accion.run();
                    dialog.dismiss();
                })
                .setNegativeButton("Cancelar", (dialog, which) -> dialog.dismiss())
                .setIcon(android.R.drawable.ic_dialog_alert)
                .show();
    }

    private void mostrarDialogoConfirmacionCompleto() {
        new AlertDialog.Builder(this)
                .setTitle("⚠️ REINICIO COMPLETO")
                .setMessage("ATENCIÓN: Esta acción eliminará TODOS los datos:\n\n" +
                        "• Inventario del sistema\n" +
                        "• Inventario escaneado\n" +
                        "• Resultados de comparación\n" +
                        "• Datos guardados en el dispositivo\n\n" +
                        "Esta acción NO se puede deshacer.\n\n" +
                        "¿Estás completamente seguro?")
                .setPositiveButton("SÍ, ELIMINAR TODO", (dialog, which) -> {
                    ejecutarReinicioCompleto();
                    dialog.dismiss();
                })
                .setNegativeButton("Cancelar", (dialog, which) -> dialog.dismiss())
                .setIcon(android.R.drawable.ic_dialog_alert)
                .show();
    }

    private void ejecutarReinicioComparacion() {
        viewModel.reiniciarComparacion();
        guardarEnHistorial("Reinicio de comparación");
        Toast.makeText(this, "Comparación reiniciada exitosamente", Toast.LENGTH_LONG).show();
        actualizarEstadoSistema();
        cargarHistorialReinicios();
    }

    private void ejecutarReinicioEscaneado() {
        viewModel.reiniciarInventarioEscaneado();
        guardarEnHistorial("Reinicio de inventario escaneado");
        Toast.makeText(this, "Inventario escaneado reiniciado exitosamente", Toast.LENGTH_LONG).show();
        actualizarEstadoSistema();
        cargarHistorialReinicios();
    }

    private void ejecutarReinicioCompleto() {
        viewModel.reiniciarCompletamente();
        guardarEnHistorial("Reinicio completo del sistema");
        Toast.makeText(this, "Reinicio completo realizado. Todos los datos han sido eliminados.", Toast.LENGTH_LONG).show();
        actualizarEstadoSistema();
        cargarHistorialReinicios();
    }

    private void guardarEnHistorial(String accion) {
        String timestamp = new SimpleDateFormat("dd/MM/yyyy HH:mm", Locale.getDefault()).format(new Date());
        String entrada = timestamp + " - " + accion;

        // Obtener historial actual
        String historialActual = sharedPreferences.getString("historial", "");

        // Agregar nueva entrada al principio
        String nuevoHistorial = entrada + "\n" + historialActual;

        // Limitar a las últimas 10 entradas
        String[] lineas = nuevoHistorial.split("\n");
        StringBuilder historialLimitado = new StringBuilder();
        for (int i = 0; i < Math.min(10, lineas.length); i++) {
            if (!lineas[i].trim().isEmpty()) {
                historialLimitado.append("• ").append(lineas[i]).append("\n");
            }
        }

        // Guardar en SharedPreferences
        sharedPreferences.edit()
                .putString("historial", nuevoHistorial)
                .apply();
    }

    private void cargarHistorialReinicios() {
        String historial = sharedPreferences.getString("historial", "");

        if (historial.isEmpty()) {
            tvHistorialReinicios.setText("No hay reinicios registrados");
        } else {
            String[] lineas = historial.split("\n");
            StringBuilder historialMostrar = new StringBuilder();
            for (int i = 0; i < Math.min(5, lineas.length); i++) {
                if (!lineas[i].trim().isEmpty()) {
                    historialMostrar.append("• ").append(lineas[i]).append("\n");
                }
            }
            tvHistorialReinicios.setText(historialMostrar.toString().trim());
        }
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        // Manejar navegación del drawer
        drawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }
}

File: app/src/main/java/com/example/escaneodematerialeskof/ui/reinicio/GestionReinicioActivity.kt
package com.example.escaneodematerialeskof.ui.reinicio

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.widget.TextView
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.ViewModelProvider
import com.example.escaneodematerialeskof.R
import com.example.escaneodematerialeskof.viewmodel.InventoryComparisonViewModel
import com.google.android.material.card.MaterialCardView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.navigation.NavigationView
import java.text.SimpleDateFormat
import java.util.*

class GestionReinicioActivity : AppCompatActivity() {

    private lateinit var drawerLayout: DrawerLayout
    private lateinit var navView: NavigationView
    private lateinit var tvEstadoSistema: TextView
    private lateinit var tvUltimoReinicio: TextView
    private lateinit var tvInventariosActivos: TextView
    private lateinit var tvHistorialReinicios: TextView // Cambiar de RecyclerView a TextView

    private lateinit var cardReiniciarComparacion: MaterialCardView
    private lateinit var cardReiniciarInventario: MaterialCardView
    private lateinit var cardLimpiarDatos: MaterialCardView
    private lateinit var cardReinicioCompleto: MaterialCardView

    private lateinit var inventoryComparisonViewModel: InventoryComparisonViewModel
    private lateinit var sharedPreferences: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_gestion_reinicio)

        // Inicializar ViewModel y SharedPreferences
        inventoryComparisonViewModel = ViewModelProvider(this)[InventoryComparisonViewModel::class.java]
        sharedPreferences = getSharedPreferences("app_config", Context.MODE_PRIVATE)

        initializeViews()
        setupNavigationDrawer()
        setupClickListeners()
        updateSystemInfo()

        // Manejo de back con OnBackPressedDispatcher
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
                    drawerLayout.closeDrawer(GravityCompat.START)
                } else {
                    isEnabled = false
                    onBackPressedDispatcher.onBackPressed()
                }
            }
        })
    }

    private fun initializeViews() {
        drawerLayout = findViewById(R.id.drawer_layout)
        navView = findViewById(R.id.nav_view)
        tvEstadoSistema = findViewById(R.id.tv_estado_sistema)
        tvUltimoReinicio = findViewById(R.id.tv_estado_escaneado) // Reutilizar un TextView existente
        tvInventariosActivos = findViewById(R.id.tv_estado_comparacion) // Reutilizar un TextView existente
        tvHistorialReinicios = findViewById(R.id.tv_historial_reinicios) // Cambiar tipo a TextView

        cardReiniciarComparacion = findViewById(R.id.card_reiniciar_comparacion)
        cardReiniciarInventario = findViewById(R.id.card_reiniciar_escaneado)
        cardLimpiarDatos = findViewById(R.id.card_reiniciar_completo) // Reutilizar una card existente
        cardReinicioCompleto = findViewById(R.id.card_reiniciar_completo)

        // Configurar toolbar
        val toolbar = findViewById<androidx.appcompat.widget.Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeAsUpIndicator(R.drawable.ic_menu)
    }

    private fun setupNavigationDrawer() {
        val toggle = ActionBarDrawerToggle(
            this, drawerLayout, R.string.navigation_drawer_open, R.string.navigation_drawer_close
        )
        drawerLayout.addDrawerListener(toggle)
        toggle.syncState()

        navView.setNavigationItemSelectedListener { menuItem ->
            // Manejar navegación aquí
            drawerLayout.closeDrawer(GravityCompat.START)
            true
        }
    }

    private fun setupClickListeners() {
        cardReiniciarComparacion.setOnClickListener {
            mostrarDialogoConfirmacion(
                titulo = getString(R.string.reinicio_comparacion),
                mensaje = getString(R.string.descripcion_reinicio_comparacion),
                accion = TipoReinicio.COMPARACION
            )
        }

        cardReiniciarInventario.setOnClickListener {
            mostrarDialogoConfirmacion(
                titulo = "📦 Reiniciar Inventario Escaneado",
                mensaje = "¿Estás seguro de que deseas reiniciar el inventario escaneado?\n\nEsto eliminará todos los datos escaneados pero mantendrá el inventario del sistema.",
                accion = TipoReinicio.INVENTARIO
            )
        }

        cardLimpiarDatos.setOnClickListener {
            mostrarDialogoConfirmacion(
                titulo = "🗑️ Limpiar Todos los Datos",
                mensaje = "¿Estás seguro de que deseas limpiar todos los datos?\n\nEsta acción eliminará tanto el inventario del sistema como los datos escaneados.",
                accion = TipoReinicio.DATOS
            )
        }

        cardReinicioCompleto.setOnClickListener {
            mostrarDialogoConfirmacion(
                titulo = getString(R.string.reinicio_completo),
                mensaje = getString(R.string.descripcion_reinicio_completo),
                accion = TipoReinicio.COMPLETO
            )
        }
    }

    private fun updateSystemInfo() {
        // Mostrar estado del sistema
        tvEstadoSistema.text = "Sistema funcionando correctamente"

        // Último reinicio
        val ultimoReinicio = sharedPreferences.getLong("ultimo_reinicio", 0L)
        if (ultimoReinicio > 0) {
            val dateFormat = SimpleDateFormat("dd/MM/yyyy HH:mm", Locale.getDefault())
            val fechaFormateada = dateFormat.format(Date(ultimoReinicio))
            tvUltimoReinicio.text = "Último reinicio: $fechaFormateada"
        } else {
            tvUltimoReinicio.text = "Último reinicio: Nunca"
        }

        // Contar inventarios activos
        val inventariosActivos = obtenerInventariosActivos()
        tvInventariosActivos.text = "Inventarios activos: $inventariosActivos"
    }

    private fun obtenerInventariosActivos(): Int {
        // En una implementación real, esto contaría los inventarios activos
        // Por ahora, simulamos un número basado en preferencias
        val hayInventarioSistema = sharedPreferences.getBoolean("inventario_sistema_cargado", false)
        val hayInventarioEscaneado = sharedPreferences.getBoolean("inventario_escaneado_activo", false)

        var count = 0
        if (hayInventarioSistema) count++
        if (hayInventarioEscaneado) count++

        return count
    }

    private fun mostrarDialogoConfirmacion(titulo: String, mensaje: String, accion: TipoReinicio) {
        MaterialAlertDialogBuilder(this)
            .setTitle(titulo)
            .setMessage(mensaje)
            .setPositiveButton(getString(R.string.aceptar)) { _, _ ->
                ejecutarReinicio(accion)
            }
            .setNegativeButton(getString(R.string.cancelar), null)
            .show()
    }

    private fun ejecutarReinicio(tipo: TipoReinicio) {
        try {
            when (tipo) {
                TipoReinicio.COMPARACION -> {
                    // Reiniciar solo datos de comparación
                    inventoryComparisonViewModel.clearComparisonData()
                    sharedPreferences.edit().putBoolean("comparacion_activa", false).apply()
                }
                TipoReinicio.INVENTARIO -> {
                    // Reiniciar inventario escaneado
                    inventoryComparisonViewModel.clearScannedInventory()
                    sharedPreferences.edit().putBoolean("inventario_escaneado_activo", false).apply()
                }
                TipoReinicio.DATOS -> {
                    // Limpiar todos los datos pero mantener configuración
                    inventoryComparisonViewModel.clearAllInventoryData()
                    sharedPreferences.edit()
                        .putBoolean("inventario_sistema_cargado", false)
                        .putBoolean("inventario_escaneado_activo", false)
                        .putBoolean("comparacion_activa", false)
                        .apply()
                }
                TipoReinicio.COMPLETO -> {
                    // Reinicio completo del sistema
                    inventoryComparisonViewModel.resetSystem()
                    // Resetear todas las preferencias excepto el historial
                    val historial = sharedPreferences.getString("historial_reinicios", "[]")
                    sharedPreferences.edit().clear().apply()
                    sharedPreferences.edit().putString("historial_reinicios", historial).apply()
                }
            }

            // Actualizar timestamp del último reinicio
            val ahora = System.currentTimeMillis()
            sharedPreferences.edit().putLong("ultimo_reinicio", ahora).apply()

            // Registrar en el historial
            registrarReinicio(tipo, true)

            // Actualizar UI
            updateSystemInfo()
            Toast.makeText(this, getString(R.string.reinicio_exitoso), Toast.LENGTH_SHORT).show()

        } catch (e: Exception) {
            // Registrar el error en el historial
            registrarReinicio(tipo, false)

            // Mostrar mensaje de error
            val mensajeError = getString(R.string.error_reinicio, e.message)
            Toast.makeText(this, mensajeError, Toast.LENGTH_LONG).show()
        }
    }

    private fun registrarReinicio(tipo: TipoReinicio, exito: Boolean) {
        // Crear nuevo item de historial
        val dateFormat = SimpleDateFormat("dd/MM/yyyy HH:mm", Locale.getDefault())
        val fechaActual = dateFormat.format(Date())
        val usuario = sharedPreferences.getString("nombre_usuario", "Admin") ?: "Admin"

        val tipoTexto = when (tipo) {
            TipoReinicio.COMPARACION -> "🔄"
            TipoReinicio.INVENTARIO -> "📦"
            TipoReinicio.DATOS -> "🗑️"
            TipoReinicio.COMPLETO -> "⚠️"
        }

        val tituloTexto = when (tipo) {
            TipoReinicio.COMPARACION -> "Reinicio de Comparación"
            TipoReinicio.INVENTARIO -> "Reinicio de Inventario"
            TipoReinicio.DATOS -> "Limpieza de Datos"
            TipoReinicio.COMPLETO -> "Reinicio Completo"
        }

        val nuevoItem = HistorialReinicioItem(
            tipo = tipoTexto,
            titulo = tituloTexto,
            fecha = fechaActual,
            usuario = usuario,
            exito = exito
        )

        // Obtener historial actual y agregar el nuevo
        val historialActual = obtenerHistorialReinicios().toMutableList()
        historialActual.add(0, nuevoItem) // Agregar al inicio para que el más reciente esté primero

        // Limitar a los últimos 20 registros
        val historialLimitado = if (historialActual.size > 20) {
            historialActual.subList(0, 20)
        } else {
            historialActual
        }

        // Guardar en SharedPreferences (en una implementación real usaríamos Room o Firestore)
        // Por simplicidad, convertimos a una cadena separada por delimitadores
        val historialString = historialLimitado.joinToString("|") { item ->
            "${item.tipo}~${item.titulo}~${item.fecha}~${item.usuario}~${item.exito}"
        }

        sharedPreferences.edit().putString("historial_reinicios", historialString).apply()

        // Actualizar el TextView de historial
        cargarHistorialReinicios()
    }

    private fun obtenerHistorialReinicios(): List<HistorialReinicioItem> {
        val historialString = sharedPreferences.getString("historial_reinicios", "") ?: ""
        if (historialString.isEmpty()) return emptyList()

        return historialString.split("|").mapNotNull { itemString ->
            val partes = itemString.split("~")
            if (partes.size >= 5) {
                HistorialReinicioItem(
                    tipo = partes[0],
                    titulo = partes[1],
                    fecha = partes[2],
                    usuario = partes[3],
                    exito = partes[4].toBoolean()
                )
            } else null
        }
    }

    private fun cargarHistorialReinicios() {
        val historial = obtenerHistorialReinicios()
        // Actualizar el TextView con el historial
        tvHistorialReinicios.text = historial.joinToString("\n") { "${it.fecha} - ${it.titulo} (${if (it.exito) "Éxito" else "Error"})" }
    }

    override fun onSupportNavigateUp(): Boolean {
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            drawerLayout.openDrawer(GravityCompat.START)
        }
        return true
    }
}
